src/tree/gpu_hist/feature_groups.cu

```diff
@@ -0,0 +1,71 @@
+/*!
+ * Copyright 2020 by XGBoost Contributors
+ */
+
+#include <xgboost/base.h>
+#include <algorithm>
+#include <vector>
+
+#include "feature_groups.cuh"
+
+#include "../../common/device_helpers.cuh"
+#include "../../common/hist_util.h"
+
+namespace xgboost {
+namespace tree {
+
+template <typename GradientSumT>
+void FeatureGroups::Init(const common::HistogramCuts& cuts, bool is_dense,
```

Comment:
Using a constructor and storing this object inside a unique pointer is preferable to using init functions.

My reasoning is that if the user forgets/doesn't realise the object is uninitialised you get a deferred crash - something undefined happens later in the program.

If we use a unique pointer and it is not initialised you get a crash the first time you touch it and it is more obvious what the problem is.